substrate/client/network/sync/src/lib.rs

```diff
@@ -200,45 +198,42 @@ impl Default for AllowedRequests {
 	}
 }
 
-struct SyncingMetrics {
-	pub import_queue_blocks_submitted: Counter<U64>,
-	pub import_queue_justifications_submitted: Counter<U64>,
-}
-
-impl SyncingMetrics {
-	fn register(registry: &Registry) -> Result<Self, PrometheusError> {
-		Ok(Self {
-			import_queue_blocks_submitted: register(
-				Counter::new(
-					"substrate_sync_import_queue_blocks_submitted",
-					"Number of blocks submitted to the import queue.",
-				)?,
-				registry,
-			)?,
-			import_queue_justifications_submitted: register(
-				Counter::new(
-					"substrate_sync_import_queue_justifications_submitted",
-					"Number of justifications submitted to the import queue.",
-				)?,
-				registry,
-			)?,
-		})
-	}
-}
-
 struct GapSync<B: BlockT> {
 	blocks: BlockCollection<B>,
 	best_queued_number: NumberFor<B>,
 	target: NumberFor<B>,
 }
 
-/// An event used to notify [`engine::SyncingEngine`] if we want to perform a block request
-/// or drop an obsolete pending response.
-enum BlockRequestEvent<B: BlockT> {
+/// Action that [`engine::SyncingEngine`] should perform after reporting imported blocks with
+/// [`ChainSync::on_blocks_processed`].
+enum BlockRequestAction<B: BlockT> {
```

Comment:
Ideally all these block/import/response actions would be under one `SyncAction` enum. `SyncingEngine` calls `ChainSync`, `ChainSync` pushes the actions it wants `SyncingEngine` to perform to its internal action queue and after each call to `ChainSync`, `SyncingEngine` would do `while let Some(action) = self.chain_sync.next_action() { match action ... }` until all actions are consumed. We'd get rid of the `OnBlockResponse::Nothing` for free.

Not in this PR though but a future task

Reply:
Makes sense, but may be returning actions from every call to `ChainSync` instead fetching them from `next_action()`? Otherwise we'll need to be extremely careful to call `next_action()` after every other call to `ChainSync`.

Reply:
I wouldn't go with returning an action from every function. Some calls to `ChainSync` could result in more than one action so we'd have to return an iterator of actions from every call or make exceptions that some calls return an iterator where others return an `Option<Action>` and then handle those differently on the call site. Doesn't seem too elegant.

If `ChainSync` is only called when there's an I/O-related event, the handling of actions can be a simple loop after `select!`